### ISSUE TITLE
Come up with full implementation list with time estimations per task

### ISSUE THREAD

**Issue by rachanahegde** [Vis4Sense/student-projects#85]
# MVP Full Implementation List

**Overall Time Frame: February 20 – April ??**

***RQ 1: How can an AI image generation application assist writers with overcoming writer’s block?***

***RQ 1a: How can a generative AI application support creative writers with limited AI knowledge in fine-tuning diffusion models for consistent character generation?***

# Feb 25 - Mar 10

## Feature: Generating the character art dataset (with Dall E)

- Test out prompt design strategies for generating consistent character art in chat GPT. Request GPT-4 to edit prompts in different ways and use them to generate images with Dall E 3. **Completed**
    - Determine whether I can use prompt design to get 1-3 character(s) consistent across multiple settings and how to make this happen more reliably. **Completed**
- Read OpenAI documentation for image generation with Dall E 3 and Dall E 2 **Completed**
- Design the user interface to gather the information I need for more effectively designing prompts.
    - Info needed: brief character description that has information about ethnicity, age, face shape, hairstyle, height, and other unique characteristics (i.e. tattoos, scars).
    - The user will be advised to ensure that the same subject is depicted in different poses, settings, and lighting for best results. And that 5 training images are needed.
- Application makes API call to GPT-4 which will edit user entered prompts.
- Application submits edited prompts to DALL·E 3 via API to generate 3 images for user to filter through.
- User can download images.
- User can edit images by specifying what changes should be made.
- Test whether inclusive and diverse character art can be generated with Dall E 3 and how to achieve this.

## Feature: Building character art dataset with images found online

- Connect to Unsplash image search API
    - Is there a way to help the user find images of one person or one animal (depending on use case) - like some kind of filter (i.e. checking captions of images that were uploaded by one user?). The obvious solution is for the user to visually identify whether the person is the same but there must be a way to streamline this a little bit?

# **Mar 11 - 17**

### Storyboarding

- Scene divs: drag and drop for images and text, add/remove, edit scene title and description text **Completed**
- Upload images **In progress**
- Save changes to local storage **In progress**

## Project creation page

- Users can enter information (i.e. title, genre, themes, story summary, and image style preference)
- Click button to create new storyboard
- Decide what other info I need for guiding prompting at later stages!

# **Mar 18 – 28**

### **Feature: Fine-tuning Stable Diffusion**

- Determine which method (Dreambooth, Custom Diffusion, LoRA, etc.) is most effective => test out finetuning strategies with google colab
- Try finetuning Stable Cascade (latest model from Stability AI, might be more efficient)
- Determine how to automate the finetuning process so that it can be trigged through the application (instead of manually in google colab)
    - Compare cost/benefits of using different platforms such as Amazon
- User can upload images to the website for finetuning
- Do processing of user generated images
    - Check for correct resolution + formatting
    - In the training images, the subject should be visible from different camera angles, striking various poses, and positioned in diverse settings.
- Put together dataset of class images with Unsplash - need to use this in the model training and send to model.
    - For Custom Diffusion - Make sure class images are NOT generated by the model and are REAL images.
- Figure out how to set parameters + weights for training depending on use case

**Use Case #1: Training model on dataset of specific human**

- Check Hugging Face blog post which discusses experimental results from training the model on humans.

**Use Case #2: Training model on dataset of specific animal**

**Use Case #3: Training model on multiple people**

# Mar 29 – April 10

## Feature: editing prompts for generating scenes with finetuned diffusion model

- Test out few shot prompting with GPT-4 to get better image prompts for Stable Diffusion
- Give prompt improvement suggestions to user
- User can approve suggestions
- Prompt is updated when user selects and approves suggestions
- Edit the final prompt to make sure it contains a modifier token representing each unique subject. Also append attributes of the character’s appearance (i.e. ethnicity, height, age, sex).
- Figure out how to host the finetuned SD model on website such as Amazon
- User can generate up to images per scene with stable diffusion model
- Evaluate how finetuned model performs with complex prompts
- Place a cap on how many images user can generate overall and per scene and notify them of this *calculate costs to determine this cap
- User can download images onto computer
- Evaluate whether the images:
    - Are useful as inspiration or reference
    - Immersive
    - Enable writer to improve their descriptions
    - Help with story visualisation
- During image generation, user can:
    - Control specific details of images of settings and characters. For example, the user can add specific objects to an image or traits (skin colour) to a character’s appearance.
    - Control over image style and mood
    - Control over degree of randomness of images
    - Negative prompt field
    - Advanced settings (i.e. generation steps, model, seed, prompt strength, width and height of image) can be optional/hidden unless needed but have default values set
    - Image count – generate between 1 to 3 images
    - Option to adjust positive and negative weighting for different keywords in prompt
    - Some way to fix issues with incorrect limbs or other issues with physical appearance of character in generated images ***maybe can use image 2 image pipeline for finetuned model?**
    - Make it clear why images are flagged as inappropriate and suggest alternative prompts or parameters to fix this ***will images be flagged?**
    - Make it clear how advanced settings will affect image without user having to read extensive documentation
- Image display
    - View prompt-image pairs (prompting history)/group images together by prompt to track progression

# April 11 – 16 *** not sure I actually have time for this (!!)

## Feature: Generating plot ideas with GPT-4

- The LLM will receive a prompt containing story context and an instruction to generate plot suggestions based on this information. The context is in the form of a one-line story summary provided by the user and the content in the storyboard, including image captions and scene descriptions.
    - Optionally, the context could also include prompts submitted to the image generation model. **Figure out how easy it would be to get this information.**
    - Although the context length for GPT-4 has increased, some of the content in the storyboard will be summarized if the prompt contains more than a predetermined amount of tokens.
        - Decide how many tokens is reasonable.
- GPT-4 will also be instructed to format its responses in a way that is easier to parse using Python code on the system’s backend.
- The user can select ideas to be converted to image prompts for the fine-tuned Stable Diffusion model
    - Take the plot idea and send it to GPT-4 ⇒ ask it to convert to an image prompt along with examples of good SD prompts.
    - Show the **updated** prompt in the prompt field for the image generation feature.
